### PR TITLE
Add support for XInclude

### LINKED ISSUE #2417: Allow DOMDocument::xinclude in templates
| Q | A
| --- | ---
| Bug? | no
| New Feature? | yes
| Sulu Version | N/A
| Browser Version | N/A

#### Actual Behavior

When using [xincludes](http://php.net/manual/en/domdocument.xinclude.php) in template xml the schema validation fails

#### Expected Behavior

To be able to add xincludes or some other way to import parts of a template, so my definitions can be kept DRY.

#### Steps to Reproduce
```xml
<?xml version="1.0" ?>
<template xmlns="http://schemas.sulu.io/template/template"
          xmlns:xi="http://www.w3.org/2001/XInclude"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/template-1.0.xsd">
    <key>default</key>

    <view>ClientWebsiteBundle:templates:default</view>
    <controller>SuluWebsiteBundle:Default:index</controller>
    <cacheLifetime>2400</cacheLifetime>

    <meta>
        <title lang="en">Default</title>
    </meta>
    
    <xi:include href="path/to/my/shared-template-parts.xml#xpointer(/shared/properties)"/>
<!--
...
--->
</template>
```

#### Possible Solutions
[XmlLegacyLoader.php line 85](https://github.com/sulu/sulu/blob/b6ea0dc60fca1e8680ce6f2cff737217fe8c170e/src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php#L85)
```php
​
$xmlDocument = XmlUtils::loadFile($resource, function (DOMDocument $dom) use ($schemaPath) {
	$dom->xinclude();
	
	$schemaSource = file_get_contents((string) $schemaPath);
	
	return @$dom->schemaValidateSource($schemaSource);
});

### PR DESCRIPTION
**Pull request by gisostallenberg**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #2417
| License | MIT
| Documentation PR | N/A

#### What's in this PR?

This allows the creator of a template to use xinclude to include parts of xml

#### Example Usage
```xml
<?xml version="1.0" ?>
<template xmlns="http://schemas.sulu.io/template/template"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xmlns:xi="http://www.w3.org/2001/XInclude"
          xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/template-1.0.xsd">

    <key>mytemplate</key>

<!--
...
-->
    <xi:include href='parts/mytemplate-properties.xml'/>
```
#### To Do

- [x] Create a documentation PR

### CODE REVIEW

File: src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php [hunk 1]
```diff
@@ -82,7 +82,12 @@ public function load($resource, $type = 'page')
         $schemaPath = __DIR__ . static::SCHEME_PATH;
 
         // read file
-        $xmlDocument = XmlUtils::loadFile($resource, $schemaPath);
+        $xmlDocument = XmlUtils::loadFile($resource, function (\DOMDocument $dom) use ($resource, $schemaPath) {
+                $dom->documentURI = $resource;
+                $dom->xinclude();
+
+                return @$dom->schemaValidate($schemaPath);
```
Comment (patkar) [inline src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php]: Is there a reason why you are silencing eventual errors?

**Comment by wachterjohannes** [comment]
@gisostallenberg looks quite good. could you also add a testcase for the xinclude?

File: src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php (hunk 1 above)
Comment (gisostallenberg) [inline src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php]: The errors get handled afterwards in XmlUtils::getXmlErrors, which are more detailed according to a [comment](http://php.net/manual/en/domdocument.schemavalidate.php#62032) on the php documentation.

Comment (patkar) [inline src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php]: Aahm the good old php inconsistencies ;)

**Comment by gisostallenberg** [comment]
@wachterjohannes Latest commit contains the test

**Comment by gisostallenberg** [comment]
Documentation PR in https://github.com/sulu/sulu-docs/pull/226

**Comment by gisostallenberg** [comment]
@danrot @wachterjohannes when this will be merged, in what release will this be available? I would like to use it from a stable release :grinning:

**Comment by danrot** [comment]
@gisostallenberg With the next minor release after its merge, which is after every quarter of the year.

File: tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" ?>
+<properties xmlns="http://schemas.sulu.io/template/template"
```
Comment (wachterjohannes) [inline tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml]: wouldn't use such a huch template for tests.

Comment (gisostallenberg) [inline tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml]: I do not see what size has to do with effectiveness of the test, but I'm happy to change it if you insist

Comment (wachterjohannes) [inline tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml]: the testcase should only test this little usecase wether the include works or not. everything else should be tested othwise. it is better readable if you dont have a huge array in the test-class

Comment (gisostallenberg) [inline tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml]: Ok, I'll change that to use a minimal template

Comment (wachterjohannes) [inline tests/Resources/DataFixtures/Page/template_with_xinclude_properties.xml]: thanks

**Comment by wachterjohannes** [comment]
@danrot what would you say? after make the test smaller it could be merged in my opinion!

**Comment by gisostallenberg** [comment]
@danrot @wachterjohannes The xinclude test uses less xml as template.
Is the test at https://github.com/sulu/sulu/blob/develop/src/Sulu/Bundle/CustomUrlBundle/Tests/Functional/Controller/CustomUrlControllerTest.php#L662 correct?

**Comment by wachterjohannes** [comment]
@gisostallenberg yes it is correct and until now it was no problem. i have restarted the test to see if it is "only" flaky. then we should make it stable (in another PR).

**Comment by wachterjohannes** [comment]
@gisostallenberg i nearly forgot to say (: you should add the CHANGELOG line for this feature. after that i will merge it.

**Comment by danrot** [comment]
Has anybody checked how the suggestion in IntelliJ works now? Ask based on the discussion in https://github.com/sulu/sulu/issues/2417#issuecomment-221529026. IMO in the ideal case the attributes like `xml:base`, which are only required for the import, but should not be used in the template itself, should not appear in the suggestions.

However, having this feature is definitely more important than insisting on this fixing this flaw, but if there would be a simple solution it might be worth doing it now.

**Comment by wachterjohannes** [comment]
@danrot no i have not tested it in intellij (:

**Comment by gisostallenberg** [comment]
@wachterjohannes What should I put in the CHANGELOG?
`FEATURE #2424 Add support for XInclude` ?

**Comment by wachterjohannes** [comment]
add [Content] after the pr number then its perfect (:

**Comment by gisostallenberg** [comment]
Same failure again in test,. I think this test is wrong, because it does not make a DateTime to compare at the same time the response get created.

**Comment by wachterjohannes** [comment]
@gisostallenberg your right ... maybe we should introduce a variable for that directly after request. would you do that please?

**Comment by gisostallenberg** [comment]
It passes the tests... :tada:

**Comment by wachterjohannes** [comment]
@gisostallenberg thanks for you patience and the very helpful contribution.

**Comment by danrot** [comment]
@gisostallenberg Thank you for the great feature 😃 But I am afraid to that the appveyor build fails 😕

**Comment by gisostallenberg** [comment]
@danrot I thought that was known and acceptable. I will try to fix the xincludes for Windows (probably a realpath on documentURI or something similar) and create a PR for you.

**Pull request merged**